File: Bisq/src/main/java/com/joachimneumann/bisq/Database/NotificationRepository.java
package com.joachimneumann.bisq.Database;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;

import java.util.List;

public class NotificationRepository {
    private RawBisqNotificationDao rawBisqNotificationDao;
    private LiveData<List<RawBisqNotification>> allRawBisqNotifications;

    public NotificationRepository(Application application) {
        NotificationDatabase db = NotificationDatabase.getDatabase(application);
        rawBisqNotificationDao = db.rawBisqNotificationDao();
        allRawBisqNotifications = rawBisqNotificationDao.getAll();
    }

    public LiveData<List<RawBisqNotification>> getAllRawBisqNotifications() {
        return allRawBisqNotifications;
    }

    public void insert (RawBisqNotification rawBisqNotification) {
        new insertAsyncTask(rawBisqNotificationDao).execute(rawBisqNotification);
    }

    private static class insertAsyncTask extends AsyncTask<RawBisqNotification, Void, Void> {
        private RawBisqNotificationDao mAsyncTaskDao;

        insertAsyncTask(RawBisqNotificationDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(final RawBisqNotification... params) {
            mAsyncTaskDao.insert(params[0]);
            return null;
        }
    }
}

File: Bisq/src/main/java/com/joachimneumann/bisq/database/NotificationRepository.kt
package bisq.android.database

import android.content.Context
import androidx.lifecycle.LiveData
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

class NotificationRepository(context: Context) {

    private val bisqNotificationDao: BisqNotificationDao

    val allBisqNotifications: LiveData<List<BisqNotification>>

    init {
        val db = NotificationDatabase.getDatabase(context)
        bisqNotificationDao = db.bisqNotificationDao()
        allBisqNotifications = bisqNotificationDao.all
    }

    suspend fun insert(bisqNotification: BisqNotification) = coroutineScope {
        launch { bisqNotificationDao.insert(bisqNotification) }
    }

    suspend fun delete(bisqNotification: BisqNotification) = coroutineScope {
        launch { bisqNotificationDao.delete(bisqNotification) }
    }

    fun getFromUid(uid: Int): BisqNotification? {
        var x: BisqNotification?
        runBlocking {
            coroutineScope {
                x = bisqNotificationDao.getFromUid(uid)
            }
        }
        return x
    }

    suspend fun deleteAll() = coroutineScope {
        launch { bisqNotificationDao.deleteAll() }
    }

    suspend fun markAllAsRead() = coroutineScope {
        launch { bisqNotificationDao.markAllAsRead(true) }
    }

    suspend fun markAsRead(uid: Int) = coroutineScope {
        launch { bisqNotificationDao.markAsRead(uid, true) }
    }
}
